### ISSUE TITLE
feature use cmake-formt and cmake-lint

### ISSUE THREAD

**Issue by ClausKlein** [MarkSchofield/WindowsToolchain#47]
Are you interested to use `pip install cmake-format` ?

with this `.cmake-format.yaml` file
```yaml
format:
  line_width: 175
  tab_size: 4
  max_pargs_hwrap: 3
  separate_ctrl_name_with_space: false
  separate_fn_name_with_space: false
  dangle_parens: true
  line_ending: unix
markup:
  bullet_char: "*"
  enum_char: .
  enable_markup: false
```

the result may be:
```
bash-3.2$ cmake-lint *.cmake    
Ninja.cmake
===========
Ninja.cmake:79,17: [E1126] Invalid form descriminator

NuGet.cmake
===========
NuGet.cmake:82,00: [C0112] Empty docstring on function or macro declaration
NuGet.cmake:82,31: [C0103] Invalid argument name "NUGET_PACKAGE_NAME" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:82,50: [C0103] Invalid argument name "NUGET_PACKAGE_VERSION" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:82,72: [C0103] Invalid argument name "VARIABLE_NAME" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:83,08: [C0103] Invalid local variable name "OPTIONS" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:84,08: [C0103] Invalid local variable name "ONE_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:85,08: [C0103] Invalid local variable name "MULTI_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:90,12: [C0103] Invalid local variable name "NUGET_PACKAGE_ROOT_PATH" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:93,08: [C0103] Invalid local variable name "NUGET_PACKAGE_PATH" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:96,12: [C0103] Invalid local variable name "NUGET_COMMAND" doesn't match `[a-z][a-z0-9_]+`
NuGet.cmake:117,12: [C0307] Bad indentation:
            )
        ^----BodyNode: 1:0->FlowControlNode: 82:0->BodyNode: 82:86->IfBlockNode: 95:4->BodyNode: 95:42->StatementNode: 112:8->TreeNode: 117:12


PowerShell.cmake
================
PowerShell.cmake:38,00: [C0112] Empty docstring on function or macro declaration
PowerShell.cmake:38,28: [C0103] Invalid argument name "INLINE_SCRIPT" doesn't match `[a-z][a-z0-9_]+`
PowerShell.cmake:41,08: [C0103] Invalid local variable name "OPTIONS" doesn't match `[a-z][a-z0-9_]+`
PowerShell.cmake:42,08: [C0103] Invalid local variable name "ONE_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
PowerShell.cmake:43,08: [C0103] Invalid local variable name "MULTI_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
PowerShell.cmake:51,08: [C0103] Invalid local variable name "POWERSHELL_COMMAND" doesn't match `[a-z][a-z0-9_]+`

ToolchainCommon.cmake
=====================
ToolchainCommon.cmake:37,00: [C0112] Empty docstring on function or macro declaration
ToolchainCommon.cmake:38,08: [C0103] Invalid local variable name "OPTIONS" doesn't match `[a-z][a-z0-9_]+`
ToolchainCommon.cmake:39,08: [C0103] Invalid local variable name "ONE_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
ToolchainCommon.cmake:40,08: [C0103] Invalid local variable name "MULTI_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
ToolchainCommon.cmake:68,00: [C0112] Empty docstring on function or macro declaration
ToolchainCommon.cmake:68,35: [C0103] Invalid argument name "VARIABLE_NAME" doesn't match `[a-z][a-z0-9_]+`
ToolchainCommon.cmake:95,00: [C0112] Empty docstring on function or macro declaration
ToolchainCommon.cmake:96,08: [C0103] Invalid local variable name "OPTIONS" doesn't match `[a-z][a-z0-9_]+`
ToolchainCommon.cmake:97,08: [C0103] Invalid local variable name "ONE_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
ToolchainCommon.cmake:98,08: [C0103] Invalid local variable name "MULTI_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`

VSWhere.cmake
=============
VSWhere.cmake:28,04: [C0103] Invalid directory variable name "_ProgramFiles" doesn't match `[A-Z][0-9A-Z_]+|_[0-9a-z_]+`
VSWhere.cmake:76,00: [C0112] Empty docstring on function or macro declaration
VSWhere.cmake:77,08: [C0103] Invalid local variable name "OPTIONS" doesn't match `[a-z][a-z0-9_]+`
VSWhere.cmake:78,08: [C0103] Invalid local variable name "ONE_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
VSWhere.cmake:79,08: [C0103] Invalid local variable name "MULTI_VALUE_KEYWORDS" doesn't match `[a-z][a-z0-9_]+`
VSWhere.cmake:83,08: [C0103] Invalid local variable name "VSWHERE_COMMAND" doesn't match `[a-z][a-z0-9_]+`
VSWhere.cmake:106,08: [C0307] Bad indentation:
        )
    ^----BodyNode: 1:0->FlowControlNode: 76:0->BodyNode: 76:26->StatementNode: 103:4->TreeNode: 106:8

VSWhere.cmake:110,04: [C0111] Missing docstring on function or macro declaration
VSWhere.cmake:110,32: [C0103] Invalid argument name "VSWHERE_OUTPUT" doesn't match `[a-z][a-z0-9_]+`
VSWhere.cmake:110,47: [C0103] Invalid argument name "VSWHERE_PROPERTY" doesn't match `[a-z][a-z0-9_]+`
VSWhere.cmake:110,64: [C0103] Invalid argument name "VARIABLE_NAME" doesn't match `[a-z][a-z0-9_]+`

Windows.Clang.toolchain.cmake
=============================

Windows.Kits.cmake
==================
Windows.Kits.cmake:25: [C0301] Line too long (205/175)
Windows.Kits.cmake:26: [C0301] Line too long (205/175)
Windows.Kits.cmake:27: [C0301] Line too long (205/175)
Windows.Kits.cmake:28: [C0301] Line too long (205/175)
Windows.Kits.cmake:29: [C0301] Line too long (205/175)
Windows.Kits.cmake:30: [C0301] Line too long (205/175)
Windows.Kits.cmake:31: [C0301] Line too long (205/175)

Windows.MSVC.toolchain.cmake
============================
Windows.MSVC.toolchain.cmake:138,24: [C0103] Invalid argument name "COMPILER" doesn't match `[a-z][a-z0-9_]+`
Windows.MSVC.toolchain.cmake:138,33: [C0103] Invalid argument name "MSVC_VERSION_OUTPUT" doesn't match `[a-z][a-z0-9_]+`
Windows.MSVC.toolchain.cmake:146,12: [C0103] Invalid local variable name "COMPILER_VERSION" doesn't match `[a-z][a-z0-9_]+`
Windows.MSVC.toolchain.cmake:147,12: [C0103] Invalid local variable name "COMPILER_VERSION_MAJOR" doesn't match `[a-z][a-z0-9_]+`
Windows.MSVC.toolchain.cmake:148,12: [C0103] Invalid local variable name "COMPILER_VERSION_MINOR" doesn't match `[a-z][a-z0-9_]+`

Summary
=======
files scanned: 8
found lint:
  Convention: 50
       Error: 1

bash-3.2$ 
```

**Comment by ClausKlein** [comment]
bash-3.2$ `cmake-format Ninja.cmake | diff -c Ninja.cmake -`
```diff
*** Ninja.cmake	Fri Oct 14 07:20:10 2022
--- -	Fri Oct 14 07:20:21 2022
***************
*** 46,56 ****
      endif()
  endif()
  
! find_program(NINJA_PATH
!     ninja ninja.exe
!     PATHS
!         ${NINJA_PATH}
! )
  
  include("${CMAKE_CURRENT_LIST_DIR}/ToolchainCommon.cmake")
  
--- 46,52 ----
      endif()
  endif()
  
! find_program(NINJA_PATH ninja ninja.exe PATHS ${NINJA_PATH})
  
  include("${CMAKE_CURRENT_LIST_DIR}/ToolchainCommon.cmake")
  
***************
*** 68,88 ****
      set(NINJA_ARCHIVE_PATH "${TOOLCHAIN_TOOLS_PATH}/ninja.zip")
      set(NINJA_PATH "${TOOLCHAIN_TOOLS_PATH}/ninja.exe")
  
!     if ((NOT CMAKE_MAKE_PROGRAM) OR (CMAKE_MAKE_PROGRAM STREQUAL NINJA_PATH))
          toolchain_download_file(
!             URL "https://github.com/ninja-build/ninja/releases/download/v${NINJA_VERSION}/ninja-win.zip"
!             PATH ${NINJA_ARCHIVE_PATH}
!             EXPECTED_HASH ${NINJA_ARCHIVE_HASH}
          )
  
          if(${NINJA_ARCHIVE_PATH} IS_NEWER_THAN ${NINJA_PATH})
!             file(ARCHIVE_EXTRACT
!                 INPUT ${NINJA_ARCHIVE_PATH}
!                 DESTINATION ${TOOLCHAIN_TOOLS_PATH}
              )
              file(TOUCH_NOCREATE ${NINJA_PATH})
          endif()
  
!         set(CMAKE_MAKE_PROGRAM ${NINJA_PATH} CACHE FILEPATH "" FORCE)
      endif()
  endif()
--- 64,93 ----
      set(NINJA_ARCHIVE_PATH "${TOOLCHAIN_TOOLS_PATH}/ninja.zip")
      set(NINJA_PATH "${TOOLCHAIN_TOOLS_PATH}/ninja.exe")
  
!     if((NOT CMAKE_MAKE_PROGRAM) OR (CMAKE_MAKE_PROGRAM STREQUAL NINJA_PATH))
          toolchain_download_file(
!             URL
!             "https://github.com/ninja-build/ninja/releases/download/v${NINJA_VERSION}/ninja-win.zip"
!             PATH
!             ${NINJA_ARCHIVE_PATH}
!             EXPECTED_HASH
!             ${NINJA_ARCHIVE_HASH}
          )
  
          if(${NINJA_ARCHIVE_PATH} IS_NEWER_THAN ${NINJA_PATH})
!             file(
!                 ARCHIVE_EXTRACT
!                 INPUT
!                 ${NINJA_ARCHIVE_PATH}
!                 DESTINATION
!                 ${TOOLCHAIN_TOOLS_PATH}
              )
              file(TOUCH_NOCREATE ${NINJA_PATH})
          endif()
  
!         set(CMAKE_MAKE_PROGRAM
!             ${NINJA_PATH}
!             CACHE FILEPATH "" FORCE
!         )
      endif()
  endif()
```
bash-3.2$

**Comment by ClausKlein** [comment]
You may change any config value as you want.

I use  mainly `cmake-format`.

You may also write rules for your `functions` and `macros`. i.e.:

```yaml
_help_parse: Options affecting listfile parsing
parse:
  _help_additional_commands:
  - Specify structure for custom cmake functions
  additional_commands:
    catch_discover_tests:
      pargs:
        nargs: '*'
        flags: []
      kwargs:
        OUTPUT_DIR: 1
        OUTPUT_PREFIX: 1
        OUTPUT_SUFFIX: 1
        REPORTER: 1
        TEST_PREFIX: 1
    conan_add_remote:
      pargs:
        nargs: '*'
        flags: []
      kwargs:
        INDEX: 1
        NAME: 1
        URL: 1
    conan_cmake_install:
      pargs:
        nargs: '*'
        flags: []
      kwargs:
        BUILD: +
        INSTALL_FOLDER: 1
        LOCKFILE: 1
        LOCKFILE_NODE_ID: 1
        LOCKFILE_OUT: 1
        PATH_OR_REFERENCE: 1
        REFERENCE: 1
        REMOTE: 1
        ENV: +
        ENV_BUILD: +
        ENV_HOST: +
        GENERATOR: +
        OPTIONS: +
        OPTIONS_BUILD: +
        OPTIONS_HOST: +
        PROFILE: +
        PROFILE_BUILD: +
        PROFILE_HOST: +
        SETTINGS: +
        SETTINGS_BUILD: +
        SETTINGS_HOST: +
```

**Comment by ClausKlein** [comment]
Hint: use `cmake-format -i ../.cmake-format.yaml --dump-config yaml --no-default`
to see the help for the config parameters:
```yaml
_help_format: Options affecting formatting.
format:
  _help_line_width:
  - How wide to allow formatted cmake files
  line_width: 120
  _help_tab_size:
  - How many spaces to tab for indent
  tab_size: 4
  _help_use_tabchars:
  - If true, lines are indented using tab characters (utf-8
  - 0x09) instead of <tab_size> space characters (utf-8 0x20).
  - In cases where the layout would require a fractional tab
  - character, the behavior of the  fractional indentation is
  - governed by <fractional_tab_policy>
  use_tabchars: false
  _help_max_pargs_hwrap:
  - If a positional argument group contains more than this many
  - arguments, then force it to a vertical layout.
  max_pargs_hwrap: 3
  _help_separate_ctrl_name_with_space:
  - If true, separate flow control names from their parentheses
  - with a space
  separate_ctrl_name_with_space: false
  _help_separate_fn_name_with_space:
  - If true, separate function names from parentheses with a
  - space
  separate_fn_name_with_space: false
  _help_dangle_parens:
  - If a statement is wrapped to more than one line, than dangle
  - the closing parenthesis on its own line.
  dangle_parens: true
  _help_line_ending:
  - What style line endings to use in the output.
  line_ending: windows
  _help_command_case:
  - Format command names consistently as 'lower' or 'upper' case
  command_case: lower
_help_markup: Options affecting comment reflow and formatting.
markup:
  _help_bullet_char:
  - What character to use for bulleted lists
  bullet_char: '*'
  _help_enum_char:
  - What character to use as punctuation after numerals in an
  - enumerated list
  enum_char: .
  _help_enable_markup:
  - enable comment markup parsing and reflow
  enable_markup: false
_help_lint: Options affecting the linter
lint:
  _help_local_var_pattern:
  - regular expression pattern describing valid names for
  - variables with local scope
  local_var_pattern: '[A-Z][A-Z0-9_]+'
  _help_argument_var_pattern:
  - regular expression pattern describing valid names for
  - function/macro arguments and loop variables.
  argument_var_pattern: '[A-Z][A-Z0-9_]+'
_help_encode: Options affecting file encoding
encode:
  _help_emit_byteorder_mark:
  - If true, emit the unicode byte-order mark (BOM) at the start
  - of the file
  emit_byteorder_mark: false
  _help_input_encoding:
  - Specify the encoding of the input file. Defaults to utf-8
  input_encoding: utf-8
  _help_output_encoding:
  - Specify the encoding of the output file. Defaults to utf-8.
  - Note that cmake only claims to support utf-8 so be careful
  - when using anything else
  output_encoding: utf-8
```

**Comment by MarkSchofield** [comment]
Fixed with #51
